#!/usr/bin/env Rscript



filter_gisticdata=function(data,gisticscorefile,qthreshold=0.1)
{
  geneposition=read.table(file="/fh/fast/dai_j/CancerGenomics/Ovarian_Cancer/result/geneposition_firehose.txt",header=T,sep="\t")
  genes=colnames(data)[2:ncol(data)] #first column is platinumclass
  df_genes=data.frame(gene=genes)
  genetable=merge(df_genes,geneposition)
  genetable=sortgenetable(genetable)
  gisticscore=read.table(file=gisticscorefile,header=T,sep="\t")
  ampgisticscore=gisticscore[gisticscore$Type=="Amp",]
  delgisticscore=gisticscore[gisticscore$Type=="Del",]
  idxkeep=rep(F,nrow(genetable))
  ampqvalues=delqvalues=rep(NA,nrow(genetable))
  library(GenomicRanges)
  gr_ampgisticscore=GRanges(seqnames=ampgisticscore$Chromosome,ranges=IRanges(start=ampgisticscore$Start,end=ampgisticscore$End),qvalue=10^(-ampgisticscore$X.log10.q.value.))
  gr_delgisticscore=GRanges(seqnames=delgisticscore$Chromosome,ranges=IRanges(start=delgisticscore$Start,end=delgisticscore$End),qvalue=10^(-delgisticscore$X.log10.q.value.))
  
  for (i in 1:nrow(genetable))
  {
    gr_gene=GRanges(seqnames=genetable$chr[i],ranges=IRanges(start=genetable$start[i],end=genetable$end[i]))
    olpa=subsetByOverlaps(gr_ampgisticscore,gr_gene)
    if (length(olpa)>0)
    {
      qvalue=min(mcols(olpa)$qvalue)
      ampqvalues[i]=qvalue
      if (qvalue<qthreshold)
      {
        idxkeep[i]=T
      }
    }
    
    olpd=subsetByOverlaps(gr_delgisticscore,gr_gene)
    if (length(olpd)>0)
    {
      qvalue=min(mcols(olpd)$qvalue)
      delqvalues[i]=qvalue
      if (qvalue<qthreshold)
      {
        idxkeep[i]=T
      }
    }
  }
  #genes_qvalue=data.frame(gene=genetable$gene,chr=genetable$chr,start=genetable$start,end=genetable$end,ampqvalue=ampqvalues,delqvalue=delqvalues)
  leftgenes=as.character(genetable$gene[idxkeep])
}
smallfdrgenes=read.table(file="../result/copynumber_result_fdrlessthan0.05.txt",header=T,sep="\t",stringsAsFactors = F)
quantile(smallfdrgenes$delgisticqvalue,probs=c(0.1,0.5,0.9))
#         10%          50%          90% 
#1.662604e-33 3.694290e-26 1.924421e-03 
quantile(smallfdrgenes$ampgisticqvalue,probs=c(0.01,0.5,0.99),na.rm=T)
qthreshold=0.01
sum(smallfdrgenes$delgisticqvalue<=qthreshold | smallfdrgenes$ampgisticqvalue<=qthreshold, na.rm=T)
#[1] 150
idx=which(smallfdrgenes$delgisticqvalue>qthreshold & smallfdrgenes$ampgisticqvalue>qthreshold)
smallfdrgenes[idx,1:4]
#chr15 is gone

#work on firehose gistic data
load("/fh/fast/dai_j/CancerGenomics/Ovarian_Cancer/data/platinum_classificationdata_stringent.RData")
data=formwholedataframe(data_copynumber)
data=data[,-c(1,2)]
gisticscorefile="/fh/fast/dai_j/CancerGenomics/Ovarian_Cancer/data/firehose_20160128/gdac.broadinstitute.org_OV-TP.CopyNumber_Gistic2.Level_4.2016012800.0.0/scores.gistic"
copynumber_filtergenes=filter_gisticdata(data,gisticscorefile,qthreshold=qthreshold)

data_copynumber_filtered=formwholedataframe(data_copynumber)
idx=which(colnames(data_copynumber_filtered) %in% copynumber_filtergenes)
data_copynumber_filtered=data_copynumber_filtered[,c(3,idx)]

#work on tangent data (not genedata generated by gistic)
load("/fh/fast/dai_j/CancerGenomics/Ovarian_Cancer/data/platinum_classificationdata_stringent_tangent.RData")
gisticscorefile="/fh/fast/dai_j/CancerGenomics/Tools/GISTIC/TCGA_new_rx0_conf0.99_armpeel1_brlen0.7_broad1/scores.gistic"
copynumber_tangent_filtergenes=filter_gisticdata(data=data_copynumber_tangent,gisticscorefile,qthreshold=qthreshold)
idx=which(colnames(data_copynumber_tangent) %in% copynumber_tangent_filtergenes)
data_copynumber_tangent_filtered=data_copynumber_tangent[,c(1,idx)]
idx=which(colnames(data_copynumber_tangent_gistic) %in% copynumber_tangent_filtergenes)
data_copynumber_tangent_gistic_filtered=data_copynumber_tangent_gistic[,c(1,idx)]



#work on aocs data (not genedata generated by gistic)
load("/fh/fast/dai_j/CancerGenomics/Ovarian_Cancer/data/Australia/data_aocs_copynumber.RData")
gisticscorefile="/fh/fast/dai_j/CancerGenomics/Tools/GISTIC/AOCS_rx0_conf0.99_armpeel1_brlen0.7_broad1/scores.gistic"
copynumber_aocs_filtergenes=filter_gisticdata(data=data_aocs_mean_copynumber,gisticscorefile,qthreshold=qthreshold)
idx=which(colnames(data_aocs_mean_copynumber) %in% copynumber_aocs_filtergenes)
data_aocs_copynumber_filtered=data_aocs_mean_copynumber[,c(1,idx)]
#generated by gistic
aocs_copynumber_gistic=read.table(file="/fh/fast/dai_j/CancerGenomics/Tools/GISTIC/AOCS_rx0_conf0.99_armpeel1_brlen0.7_broad1/all_data_by_genes.txt",sep="\t")

#save(data_copynumber_filtered,data_copynumber_tangent_filtered,data_copynumber_tangent_gistic_filtered,data_aocs_copynumber_filtered,file="/fh/fast/dai_j/CancerGenomics/Ovarian_Cancer/data/platinum_classificationdata_stringent_filtered.RData")
save(data_copynumber_filtered,data_copynumber_tangent_filtered,data_copynumber_tangent_gistic_filtered,data_aocs_copynumber_filtered,file="/fh/fast/dai_j/CancerGenomics/Ovarian_Cancer/data/platinum_classificationdata_stringent_filtered.q0.01.RData")


#check smallfdrgenes
smallfdrgenes=read.table(file="../result/copynumber_result_fdrlessthan0.05.txt",header=T,sep="\t",stringsAsFactors = F)
sum(copynumber_filtergenes %in% smallfdrgenes$gene)
#[1] 157
which(! smallfdrgenes$gene %in% copynumber_filtergenes)
smallfdrgenes[151,]

sum(copynumber_tangent_filtergenes %in% smallfdrgenes$gene)
#[1] 158

sum(copynumber_aocs_filtergenes %in% smallfdrgenes$gene)
#[1] 91
